pkg/apiserver/apiserver.go

```diff
@@ -36,49 +38,133 @@ import (
+		_ = godotenv.Load()
 		gin.SetMode(gin.ReleaseMode)
 	})
 
-	r := gin.New()
-	r.Use(cors.AllowAll())
-	r.Use(gin.Recovery())
-	r.Use(gzip.Gzip(gzip.BestSpeed))
-	r.Use(utils.MWHandleErrors())
+	return &Service{
+		status:            utils2.NewServiceStatus(),
+		config:            cfg,
+		newPDDataProvider: newPDDataProvider,
+		stoppedHandler:    stoppedHandler,
+	}
+}
+
+func Register(r *gin.RouterGroup, s *Service) {
+	endpoint := r.Group("/dashboard/api")
+	endpoint.Use(s.status.MWHandleStopped(gin.WrapH(s.stoppedHandler)))
+	endpoint.Any("/*any", s.handler)
+}
+
+func (s *Service) Start(ctx context.Context) error {
+	s.app = fx.New(
+		fx.Logger(utils2.NewFxPrinter()),
+		fx.Provide(
+			newAPIHandlerEngine,
+			s.provide,
+			s.newPDDataProvider,
+			dbstore.NewDBStore,
+			pd.NewEtcdClient,
+			tidb.NewForwarderConfig,
+			tidb.NewForwarder,
+			http2.NewHTTPClientWithConf,
+			user.NewAuthService,
+			foo.NewService,
+			info.NewService,
+			clusterinfo.NewService,
+			profiling.NewService,
+			logsearch.NewService,
+			statement.NewService,
+			diagnose.NewService,
+			keyvisual.NewService,
+		),
+		fx.Populate(&s.apiHandlerEngine),
+		fx.Invoke(
+			user.Register,
+			foo.Register,
+			info.Register,
+			clusterinfo.Register,
+			profiling.Register,
+			logsearch.Register,
+			statement.Register,
+			diagnose.Register,
+			keyvisual.Register,
+			// Must be at the end
+			s.status.Invoke,
+		),
+	)
 
-	endpoint := r.Group(apiPrefix)
+	if err := s.app.Err(); err != nil {
+		return err
+	}
+	if err := s.app.Start(ctx); err != nil {
+		return err
+	}
+	return nil
+}
 
-	auth := user.NewAuthService(services.TiDBForwarder)
-	auth.Register(endpoint)
+func (s *Service) Stop(ctx context.Context) error {
+	err := s.app.Stop(ctx)
+	s.apiHandlerEngine = nil
+	return err
+}
 
-	foo.NewService(config).Register(endpoint, auth)
-	info.NewService(config, services.TiDBForwarder, services.Store).Register(endpoint, auth)
-	clusterinfo.NewService(config, services.EtcdProvider, services.HTTPClient).Register(endpoint, auth)
-	profiling.NewService(config, services.Store).Register(endpoint, auth)
-	services.KeyVisual.Register(endpoint, auth)
-	logsearch.NewService(config, services.Store).Register(endpoint, auth)
-	statement.NewService(config, services.TiDBForwarder).Register(endpoint, auth)
-	diagnose.NewService(config, services.TiDBForwarder, services.Store, NewTemplate(r, "diagnose")).Register(endpoint, auth)
+func (s *Service) NewStatefulHandler(handler http.Handler) http.Handler {
+	return s.status.NewStatefulHandler(handler, s.stoppedHandler)
+}
 
-	return r
+func (s *Service) handler(c *gin.Context) {
+	s.apiHandlerEngine.HandleContext(c)
 }
 
-func NewTemplate(r *gin.Engine, name string) *template.Template {
-	return template.New(name).Funcs(r.FuncMap)
+func (s *Service) provide() *config.Config {
+	return s.config
 }
+
+var StoppedHandler = http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
```

Comment:
Should we move this handler to cmd package?

Reply:
It needs to be used in PD.